sip/sip-005-blocks-and-transactions.md

```diff
@@ -0,0 +1,1339 @@
+
+Example: `"vm-account::SP2RZRSEQHCFPHSBHJTKNWT86W6VSK51M7BCMY06Q::18"` refers to
+the nonce of account `SP2RZRSEQHCFPHSBHJTKNWT86W6VSK51M7BCMY06Q`.
+
+The STX balance is encoded as follows:
+
+* Key: the string "vm-account::", a c32check-encoded address, and the string
+  "::19"
+* Value: a serialized Clarity `UInt`
+
+Example: `"vm-account::SP2RZRSEQHCFPHSBHJTKNWT86W6VSK51M7BCMY06Q::19"` refers to
+the STX balance of account `SP2RZRSEQHCFPHSBHJTKNWT86W6VSK51M7BCMY06Q`.
+
+A fungible token balance owned by an account is encoded as follows:
+
+* Key: the string `"vm::"`, the fully-qualified contract identifier, the string `"::2::"`,
+  the name of the token as defined in its Clarity contract, the string `"::"`, and the
+c32check-encoded address of the account.
+* Value: a serialized Clarity `UInt`
+
+Example: `"vm::SP13N5TE1FBBGRZD1FCM49QDGN32WAXM2E5F8WT2G.example-contract::2::example-token::SP2RZRSEQHCFPHSBHJTKNWT86W6VSK51M7BCMY06Q"`
+refers to the balance of `example-token` -- a fungible token defined in contract `SP13N5TE1FBBGRZD1FCM49QDGN32WAXM2E5F8WT2G.example-contract` -- 
+that is owned by account `SP2RZRSEQHCFPHSBHJTKNWT86W6VSK51M7BCMY06Q`.
+
+A non-fungible token owned by an account is encoded as follows:
+
+* Key: the string `"vm::"`, the fully-qualified contract identifier, the string `"::4::"`,
+  the name of the token as defined in its Clarity contract, the string `"::"`,
+and the serialized Clarity value that represents the token.
+* Value: a serialized Clarity standard principal
+
+Example: `"vm::SP13N5TE1FBBGRZD1FCM49QDGN32WAXM2E5F8WT2G.example-contract::4::example-nft::\x02\x00\x00\x00\x0b\x68\x65\x6c\x6c\x6f\x20\x77\x6f\x72\x6c\x64"` 
+refers to the non-fungible token `"hello world"` (which has type `buff` and is
+comprised of 11 bytes), defined in Clarity contract `SP13N5TE1FBBGRZD1FCM49QDGN32WAXM2E5F8WT2G.example-contract`
+as a type of non-fungible token `example-nft`.
+
+#### (TBD) Calculating the State of a Smart Contract
+
+A smart contract's canonical encoding as a key/value pair in the smart contract
+state is as follows:
+
+* Key: The string `"vm::"`, the fully-qualified contract identifier, and the
+  string `::9`.
+* Value: TBD
+
+(TBD) Note that _other_ kinds of smart contract data (e.g., analysis data, function bodies, etc.)
+must also be indexed by Stacks peers. However, this data does not need to be included in a
+peer's commitment -- it is indexed via a secondary MARF structure.  Therefore, the representation of
+that data is _not_ specified as part of the Stacks protocol. 
+
+Example: `"vm::SP13N5TE1FBBGRZD1FCM49QDGN32WAXM2E5F8WT2G.example-contract::9"`
+refers to the contract AST created by standard principal
+`SP13N5TE1FBBGRZD1FCM49QDGN32WAXM2E5F8WT2G` and named `example-contract`.
+
+A smart contract's persistent data variables are each encoded as two key/value
+pairs: a pair describing the variable's data type, and a pair describing the
+variable's value.
+
+The data type is encoded as follows:
```

Comment:
Persistent Variable, Map, NFT, FT Data _types_ should not be committed -- they are strictly determinable from the contract code, which is committed to.